src/librustc/arena.rs

```diff
@@ -46,39 +77,130 @@ pub trait ArenaAllocatable {}
 
 impl<T: Copy> ArenaAllocatable for T {}
 
-pub trait ArenaField<'tcx>: Sized {
+pub unsafe trait ArenaField<'tcx>: Sized {
     /// Returns a specific arena to allocate from.
-    fn arena<'a>(arena: &'a Arena<'tcx>) -> &'a TypedArena<Self>;
+    /// If None is returned, the DropArena will be used.
+    fn arena<'a>(arena: &'a Arena<'tcx>) -> Option<&'a TypedArena<Self>>;
 }
 
-impl<'tcx, T> ArenaField<'tcx> for T {
+unsafe impl<'tcx, T> ArenaField<'tcx> for T {
     #[inline]
-    default fn arena<'a>(_: &'a Arena<'tcx>) -> &'a TypedArena<Self> {
+    default fn arena<'a>(_: &'a Arena<'tcx>) -> Option<&'a TypedArena<Self>> {
         panic!()
     }
 }
 
 impl<'tcx> Arena<'tcx> {
     #[inline]
     pub fn alloc<T: ArenaAllocatable>(&self, value: T) -> &mut T {
-        if mem::needs_drop::<T>() {
-            <T as ArenaField<'tcx>>::arena(self).alloc(value)
-        } else {
-            self.dropless.alloc(value)
+        if !mem::needs_drop::<T>() {
+            return self.dropless.alloc(value);
+        }
+        match <T as ArenaField<'tcx>>::arena(self) {
+            Some(arena) => arena.alloc(value),
+            None => unsafe { self.drop.alloc(value) },
         }
     }
 
     pub fn alloc_from_iter<
         T: ArenaAllocatable,
         I: IntoIterator<Item = T>
     >(
-        &self,
+        &'a self,
         iter: I
-    ) -> &mut [T] {
-        if mem::needs_drop::<T>() {
-            <T as ArenaField<'tcx>>::arena(self).alloc_from_iter(iter)
-        } else {
-            self.dropless.alloc_from_iter(iter)
+    ) -> &'a mut [T] {
+        if !mem::needs_drop::<T>() {
+            return self.dropless.alloc_from_iter(iter);
+        }
+        match <T as ArenaField<'tcx>>::arena(self) {
+            Some(arena) => arena.alloc_from_iter(iter),
+            None => unsafe { self.drop.alloc_from_iter(iter) },
+        }
+    }
+}
+
+/// Calls the destructor for an object when dropped.
+struct DropType {
+    drop_fn: unsafe fn(*mut u8),
+    obj: *mut u8,
+}
+
+unsafe fn drop_for_type<T>(to_drop: *mut u8) {
+    std::ptr::drop_in_place(to_drop as *mut T)
+}
+
+impl Drop for DropType {
+    fn drop(&mut self) {
+        unsafe {
+            (self.drop_fn)(self.obj)
+        }
+    }
+}
+
+/// An arena which can be used to allocate any type.
+/// Allocating in this arena is unsafe since the type system
+/// doesn't know which types it contains. In order to
+/// allocate safetly, you must store a PhantomData<T>
+/// alongside this arena for each type T you allocate.
+#[derive(Default)]
+struct DropArena {
+    /// A list of destructors to run when the arena drops.
+    /// Ordered so `destructors` gets dropped before the arena
+    /// since its destructor can reference memory in the arena.
+    destructors: RefCell<Vec<DropType>>,
```

Comment:
Looking at the impl, I think this will work for values which refer to other values in the same arena, as long as those types don't have interior mutability and can end up with references to values created later.

How are we ensuring this invariant?

Reply:
That's not an invariant that's upheld. We just let dropck ensure that types can't do anything bad in their destructors.

Reply:
I'm talking about something like

```rust
struct A<'tcx> {
    x: Mutex<Option<&'tcx B>>,
}

impl<'tcx> Drop for A<'tcx> {
    fn drop(&mut self) {
        println!("{}", x.lock().unwrap().unwrap().s);
    }
}

struct B {
    s: String,
}
```

where 

1. intern a value of type `A`
2. intern a value of type `B`
3. change the value from 1. to point to 2.
4. value from 2. is dropped
5. value from 1. is dropped and accesses the value from 2. which has already been dropped

Reply:
You can the same interning with `(TypedArena<A<'tcx>>, TypedArena<B>)`.